### PR TITLE
Warn for variables/fields declared with generic type without `?`

### PR DESCRIPTION
**Pull request by mppf**
Resolves https://github.com/Cray/chapel-private/issues/5032

This PR adds a warning for these cases:
* `var x: someGenericType` for fields and variables (need `someGenericType(?)`)
* `proc foo(type t) { var x: t; }` -- similarly to `var x: someGenericType`
* `var x: f();` where `f()` returns a generic type, for fields and variables

In the above, `someGenericType` includes `domain` or user records/classes that are generic. Note that PR #22697 added a warning for fields declared with type `domain` or fields of class type with generic management.

Reviewed by @vasslitvinov - thanks!

- [x] full comm=none testing

Future work
* warn for `class C : SomeGenericType` -- PR #22784

### CODE REVIEW

File: test/studies/adventOfCode/2021/bradc/futures/day18-bug2.bad [hunk 1]
```diff
@@ -1,5 +1,7 @@
-$CHPL_HOME/modules/internal/OwnedObject.chpl:167: In initializer:
-$CHPL_HOME/modules/internal/OwnedObject.chpl:172: error: Could not coerce 'nil' to 'borrowed Node?' in initialization
-  $CHPL_HOME/modules/internal/OwnedObject.chpl:247: called as owned.init(type chpl_t = borrowed Node?)
+day18-bug2.chpl:26: warning: please use '?' when declaring a variable with generic type
+day18-bug2.chpl:26: note: for example with 'Node(?)'
+$CHPL_HOME/modules/internal/OwnedObject.chpl:50: In initializer:
+$CHPL_HOME/modules/internal/OwnedObject.chpl:55: error: Could not coerce 'nil' to 'borrowed Node?' in initialization
```
Comment (lydia-duncan) [inline test/studies/adventOfCode/2021/bradc/futures/day18-bug2.bad:4]: This is a pretty significant line difference for what looks like the same call.  Was there a change to OwnedObject.chpl that didn't get committed?

Comment (mppf) [reply test/studies/adventOfCode/2021/bradc/futures/day18-bug2.bad:4]: Nope. Line numbers are prediff'd away in .bad files so it just represents me updating it with current line numbers from other changes. But the line numbers will be ignored.

File: test/functions/generic/genericChannelInInit2.good [hunk 2]
```diff
@@ -1,3 +1,5 @@
+genericChannelInInit2.chpl:7: warning: please use '?' when declaring a field with generic type
+genericChannelInInit2.chpl:7: note: for example with 'fileWriter(?)'
```
Comment (bradcray) [inline test/functions/generic/genericChannelInInit2.good:2]: @mppf: Mostly rhetorical question for me, but may be of greater interest to a reviewer:  How did you decide whether to keep a test working by updating its source vs. its .good/.bad file?

File: compiler/resolution/functionResolution.cpp [hunk 3]
```diff
@@ -13776,44 +13779,170 @@ void checkDuplicateDecorators(Type* decorator, Type* decorated, Expr* ctx) {
...
-  if (def == nullptr) {
+static bool isBuiltinGenericType(Type* t) {
+  return isBuiltinGenericClassType(t) ||
+         t == dtAnyComplex || t == dtAnyImag || t == dtAnyReal ||
+         t == dtAnyBool || t == dtAnyEnumerated ||
+         t == dtNumeric || t == dtIntegral ||
+         t == dtIteratorRecord || t == dtIteratorClass ||
+         t == dtAnyPOD ||
+         t == dtOwned || t == dtShared ||
+         t == dtAnyRecord;
```
Comment (vasslitvinov) [inline compiler/resolution/functionResolution.cpp:13790]: Please double-check whether the above needs dtBorrowed, dtUnmanaged, dtAnyManagementAnyNilable, dtAnyNilable(??).

**Review by vasslitvinov — APPROVED** [review]
Looks good. Please see my comment.

File: compiler/resolution/functionResolution.cpp (hunk 3 above)
Comment (mppf) [reply compiler/resolution/functionResolution.cpp:13790]: These are covered by `isBuiltinGenericClassType` which is called as the first option in the `||` clauses.

**Pull request merged**